src/lib/y2storage/proposal.rb

```diff
@@ -78,91 +78,94 @@ def propose
       raise UnexpectedCallError if proposed?
       settings.freeze
       @proposed = true
-      @volumes = volumes_list(:all, populated_settings)
-      @devices = devicegraph(@volumes, populated_settings)
+
+      # FIXME: The current implementation, tries :desired and then :min for
+      # every root disk. It would probably make more sense to try first :desired
+      # for all possible root disks and then do the same with :min.
+      exception = nil
+      candidate_roots.each do |disk_name|
+        populated_settings.root_device = disk_name
+        exception = nil
+
+        begin
+          @volumes = volumes_list
+          @devices = devicegraph(@volumes)
+        rescue Error => error
+          log.info "Failed to make a proposal using root device #{disk_name}: #{error.message}"
+          exception = error
+        end
+
+        break unless exception
```

Comment:
If we have to make a proposal and we have several possible candidates, we try with every one of them, capturing the exception if anything went wrong in order to try with the next one. I find it pretty natural.

The calculation of a proposal for a particular disk can go wrong for many reasons and at many different depth levels, so exceptions are the less cumbersome way to handle that. Otherwise, every single component would need an extra mechanism to communicate the possible errors, which sounds like re-inventing exceptions.